File: frontend/src/lib/api/feedbackApproval.ts
import { apiGet, apiPost, apiPut } from '../apiClient';

// 型定義
export interface ApprovalRequest {
  analysis_id: number;
  visibility_level: 'private' | 'team' | 'organization' | 'public';
  request_reason?: string;
  is_staged_publication: boolean;
  publication_stages?: PublicationStage[];
}

export interface PublicationStage {
  stage_number: number;
  visibility_level: 'private' | 'team' | 'organization' | 'public';
  description: string;
  delay_days: number;
  auto_advance: boolean;
}

export interface FeedbackApprovalUpdate {
  approval_status?: 'pending' | 'under_review' | 'approved' | 'rejected' | 'requires_changes';
  review_notes?: string;
  rejection_reason?: string;
  visibility_level?: 'private' | 'team' | 'organization' | 'public';
}

export interface UserConfirmationRequest {
  approval_id: number;
  confirm: boolean;
  confirmation_notes?: string;
}

export interface FeedbackApprovalResponse {
  id: number;
  analysis_id: number;
  requester_id: number;
  reviewer_id?: number;
  approval_status: 'pending' | 'under_review' | 'approved' | 'rejected' | 'requires_changes';
  visibility_level: 'private' | 'team' | 'organization' | 'public';
  request_reason?: string;
  review_notes?: string;
  rejection_reason?: string;
  is_staged_publication: boolean;
  publication_stages?: string;
  current_stage: number;
  requires_confirmation: boolean;
  is_confirmed: boolean;
  confirmation_date?: string;
  requested_at: string;
  reviewed_at?: string;
  published_at?: string;
  requester_name?: string;
  reviewer_name?: string;
  analysis_title?: string;
}

export interface FeedbackApprovalListResponse {
  approvals: FeedbackApprovalResponse[];
  total_count: number;
  page: number;
  page_size: number;
}

export interface FeedbackApprovalStats {
  total_approvals: number;
  pending_approvals: number;
  approved_approvals: number;
  rejected_approvals: number;
  under_review_approvals: number;
  requires_changes_approvals: number;
  average_approval_time_hours: number;
  approval_rate: number;
  visibility_distribution: Record<string, number>;
  status_distribution: Record<string, number>;
}

export interface FeedbackApprovalFilters {
  analysis_id?: number;
  approval_status?: string;
  visibility_level?: string;
  is_confirmed?: boolean;
}

// API関数
export const feedbackApprovalApi = {
  // 承認リクエストを作成
  createApprovalRequest: async (data: ApprovalRequest): Promise<FeedbackApprovalResponse> => {
    const response = await apiPost<FeedbackApprovalResponse>('/feedback-approvals/', data);
    return response;
  },

  // 自分の承認リクエスト一覧を取得
  getMyApprovals: async (
    page: number = 1,
    pageSize: number = 20,
    filters?: FeedbackApprovalFilters
  ): Promise<FeedbackApprovalListResponse> => {
    const params = new URLSearchParams({
      page: page.toString(),
      page_size: pageSize.toString(),
    });

    if (filters?.analysis_id) {
      params.append('analysis_id', filters.analysis_id.toString());
    }
    if (filters?.approval_status) {
      params.append('approval_status', filters.approval_status);
    }
    if (filters?.visibility_level) {
      params.append('visibility_level', filters.visibility_level);
    }
    if (filters?.is_confirmed !== undefined) {
      params.append('is_confirmed', filters.is_confirmed.toString());
    }

    const response = await apiGet<FeedbackApprovalListResponse>(`/feedback-approvals/my?${params.toString()}`);
    return response;
  },

  // レビュー待ちの承認リクエスト一覧を取得
  getPendingApprovals: async (
    page: number = 1,
    pageSize: number = 20
  ): Promise<FeedbackApprovalListResponse> => {
    const params = new URLSearchParams({
      page: page.toString(),
      page_size: pageSize.toString(),
    });

    const response = await apiGet<FeedbackApprovalListResponse>(`/feedback-approvals/pending?${params.toString()}`);
    return response;
  },

  // 承認リクエストをレビュー
  reviewApproval: async (
    approvalId: number,
    data: FeedbackApprovalUpdate
  ): Promise<FeedbackApprovalResponse> => {
    const response = await apiPut<FeedbackApprovalResponse>(`/feedback-approvals/${approvalId}/review`, data);
    return response;
  },

  // ユーザーが承認リクエストを確認
  confirmApproval: async (
    approvalId: number,
    data: UserConfirmationRequest
  ): Promise<FeedbackApprovalResponse> => {
    const response = await apiPost<FeedbackApprovalResponse>(`/feedback-approvals/${approvalId}/confirm`, data);
    return response;
  },

  // 承認済みの分析結果を公開
  publishApprovedAnalysis: async (approvalId: number): Promise<{ message: string }> => {
    const response = await apiPost<{ message: string }>(`/feedback-approvals/${approvalId}/publish`, {});
    return response;
  },

  // 段階的公開の次の段階に進む
  advancePublicationStage: async (approvalId: number): Promise<{ message: string }> => {
    const response = await apiPost<{ message: string }>(`/feedback-approvals/${approvalId}/advance-stage`, {});
    return response;
  },

  // 承認統計を取得
  getApprovalStats: async (): Promise<FeedbackApprovalStats> => {
    const response = await apiGet<FeedbackApprovalStats>('/feedback-approvals/stats');
    return response;
  },

  // 承認リクエストを削除
  deleteApproval: async (approvalId: number): Promise<{ message: string }> => {
    // apiDeleteが存在しないため、fetchを使用
    const token = await getAuthToken();
    const url = `${process.env.NEXT_PUBLIC_API_BASE_URL || 'http://localhost:8000'}/api/v1/feedback-approvals/${approvalId}`;
    
    const response = await fetch(url, {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
    });

    if (!response.ok) {
      throw new Error(`DELETE /feedback-approvals/${approvalId} failed: ${response.status}`);
    }

    return response.json();
  },
};

// ヘルパー関数
async function getAuthToken(): Promise<string> {
  // 簡易的な実装（実際の実装では適切な認証処理が必要）
  return 'dummy-token';
}

// ユーティリティ関数
export const getApprovalStatusLabel = (status: string): string => {
  const statusLabels: Record<string, string> = {
    pending: '承認待ち',
    under_review: 'レビュー中',
    approved: '承認済み',
    rejected: '却下',
    requires_changes: '修正要求',
  };
  return statusLabels[status] || status;
};

export const getApprovalStatusColor = (status: string): string => {
  const statusColors: Record<string, string> = {
    pending: 'bg-yellow-100 text-yellow-800',
    under_review: 'bg-blue-100 text-blue-800',
    approved: 'bg-green-100 text-green-800',
    rejected: 'bg-red-100 text-red-800',
    requires_changes: 'bg-orange-100 text-orange-800',
  };
  return statusColors[status] || 'bg-gray-100 text-gray-800';
};

export const getVisibilityLevelLabel = (level: string): string => {
  const levelLabels: Record<string, string> = {
    private: '本人のみ',
    team: 'チーム内',
    organization: '組織内',
    public: '公開',
  };
  return levelLabels[level] || level;
};

export const getVisibilityLevelColor = (level: string): string => {
  const levelColors: Record<string, string> = {
    private: 'bg-gray-100 text-gray-800',
    team: 'bg-blue-100 text-blue-800',
    organization: 'bg-purple-100 text-purple-800',
    public: 'bg-green-100 text-green-800',
  };
  return levelColors[level] || 'bg-gray-100 text-gray-800';
};

export const formatApprovalDate = (dateString: string): string => {
  const date = new Date(dateString);
  return date.toLocaleDateString('ja-JP', {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
  });
};

export const calculateApprovalTime = (requestedAt: string, reviewedAt?: string): string => {
  if (!reviewedAt) return '-';
  
  const requested = new Date(requestedAt);
  const reviewed = new Date(reviewedAt);
  const diffMs = reviewed.getTime() - requested.getTime();
  const diffHours = Math.round(diffMs / (1000 * 60 * 60));
  
  if (diffHours < 24) {
    return `${diffHours}時間`;
  } else {
    const diffDays = Math.round(diffHours / 24);
    return `${diffDays}日`;
  }
};

File: backend/app/schemas/feedback_approval.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


class ApprovalStatus(str, Enum):
    """承認ステータス"""
    PENDING = "pending"
    UNDER_REVIEW = "under_review"
    APPROVED = "approved"
    REJECTED = "rejected"
    REQUIRES_CHANGES = "requires_changes"


class VisibilityLevel(str, Enum):
    """可視性レベル"""
    PRIVATE = "private"
    TEAM = "team"
    ORGANIZATION = "organization"
    PUBLIC = "public"


class FeedbackApprovalBase(BaseModel):
    """フィードバック承認の基本スキーマ"""
    analysis_id: int = Field(..., description="分析ID")
    visibility_level: VisibilityLevel = Field(..., description="可視性レベル")
    request_reason: Optional[str] = Field(None, description="申請理由")
    is_staged_publication: bool = Field(False, description="段階的公開フラグ")
    publication_stages: Optional[str] = Field(None, description="公開段階（JSON文字列）")
    requires_confirmation: bool = Field(True, description="本人確認が必要か")


class FeedbackApprovalCreate(FeedbackApprovalBase):
    """フィードバック承認作成用スキーマ"""
    pass


class FeedbackApprovalUpdate(BaseModel):
    """フィードバック承認更新用スキーマ"""
    approval_status: Optional[ApprovalStatus] = Field(None, description="承認ステータス")
    review_notes: Optional[str] = Field(None, description="レビューコメント")
    rejection_reason: Optional[str] = Field(None, description="却下理由")
    visibility_level: Optional[VisibilityLevel] = Field(None, description="可視性レベル")


class FeedbackApprovalResponse(FeedbackApprovalBase):
    """フィードバック承認応答用スキーマ"""
    id: int
    requester_id: int
    reviewer_id: Optional[int]
    approval_status: ApprovalStatus
    current_stage: int
    is_confirmed: bool
    confirmation_date: Optional[datetime]
    requested_at: datetime
    reviewed_at: Optional[datetime]
    published_at: Optional[datetime]
    
    # 関連情報
    requester_name: Optional[str] = None
    reviewer_name: Optional[str] = None
    analysis_title: Optional[str] = None
    
    class Config:
        from_attributes = True


class FeedbackApprovalListResponse(BaseModel):
    """フィードバック承認一覧応答用スキーマ"""
    approvals: List[FeedbackApprovalResponse]
    total_count: int
    page: int
    page_size: int


class FeedbackApprovalFilters(BaseModel):
    """フィードバック承認フィルター用スキーマ"""
    analysis_id: Optional[int] = Field(None, description="分析ID")
    requester_id: Optional[int] = Field(None, description="申請者ID")
    reviewer_id: Optional[int] = Field(None, description="レビュアーID")
    approval_status: Optional[ApprovalStatus] = Field(None, description="承認ステータス")
    visibility_level: Optional[VisibilityLevel] = Field(None, description="可視性レベル")
    is_confirmed: Optional[bool] = Field(None, description="本人確認済みか")


class FeedbackApprovalQueryParams(BaseModel):
    """フィードバック承認クエリパラメータ用スキーマ"""
    page: int = Field(default=1, ge=1, description="ページ番号")
    page_size: int = Field(default=20, ge=1, le=100, description="ページサイズ")
    analysis_id: Optional[int] = Field(None, description="分析ID")
    requester_id: Optional[int] = Field(None, description="申請者ID")
    reviewer_id: Optional[int] = Field(None, description="レビュアーID")
    approval_status: Optional[ApprovalStatus] = Field(None, description="承認ステータス")
    visibility_level: Optional[VisibilityLevel] = Field(None, description="可視性レベル")
    is_confirmed: Optional[bool] = Field(None, description="本人確認済みか")


class ApprovalRequest(BaseModel):
    """承認リクエスト用スキーマ"""
    analysis_id: int = Field(..., description="分析ID")
    visibility_level: VisibilityLevel = Field(..., description="可視性レベル")
    request_reason: Optional[str] = Field(None, description="申請理由")
    is_staged_publication: bool = Field(False, description="段階的公開フラグ")
    publication_stages: Optional[List[Dict[str, Any]]] = Field(None, description="公開段階の詳細")


class ApprovalResponse(BaseModel):
    """承認レスポンス用スキーマ"""
    approval_status: ApprovalStatus
    review_notes: Optional[str] = None
    rejection_reason: Optional[str] = None


class UserConfirmationRequest(BaseModel):
    """本人確認リクエスト用スキーマ"""
    approval_id: int = Field(..., description="承認ID")
    confirm: bool = Field(..., description="確認フラグ")
    confirmation_notes: Optional[str] = Field(None, description="確認コメント")


class PublicationStage(BaseModel):
    """公開段階の詳細スキーマ"""
    stage_number: int = Field(..., description="段階番号")
    visibility_level: VisibilityLevel = Field(..., description="可視性レベル")
    description: str = Field(..., description="段階の説明")
    delay_days: int = Field(0, description="前段階からの遅延日数")
    auto_advance: bool = Field(False, description="自動進行フラグ")


class StagedPublicationRequest(BaseModel):
    """段階的公開リクエスト用スキーマ"""
    analysis_id: int = Field(..., description="分析ID")
    stages: List[PublicationStage] = Field(..., description="公開段階のリスト")
    request_reason: Optional[str] = Field(None, description="申請理由")


class FeedbackApprovalStats(BaseModel):
    """フィードバック承認統計用スキーマ"""
    total_approvals: int
    pending_approvals: int
    approved_approvals: int
    rejected_approvals: int
    under_review_approvals: int
    requires_changes_approvals: int
    average_approval_time_hours: float
    approval_rate: float  # 承認率（0.0-1.0）
    visibility_distribution: Dict[str, int]
    status_distribution: Dict[str, int]
